libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -137,13 +137,49 @@ bool AP_Mission::starts_with_takeoff_cmd()
         cmd_index = AP_MISSION_FIRST_REAL_COMMAND;
     }
 
-    if (!get_next_nav_cmd(cmd_index, cmd)) {
-        return false;
-    }
-    if (cmd.id != MAV_CMD_NAV_TAKEOFF) {
-        return false;
+    // check a maximum of 16 items, remembering that missions can have
+    // loops in them
+    for (uint8_t i=0; i<16; i++, cmd_index++) {
+        if (!get_next_cmd(cmd_index, cmd, false)) {
+            return false;
+        }
+        switch (cmd.id) {
+        // any of these are considered a takeoff command:
+        case MAV_CMD_NAV_TAKEOFF:
+        case MAV_CMD_NAV_TAKEOFF_LOCAL:
+            return true;
+        // any of these are considered "skippable" when determining if
+        // we "start with a takeoff command"
+        case MAV_CMD_NAV_DELAY:
+        case MAV_CMD_DO_CHANGE_SPEED:
+        case MAV_CMD_DO_JUMP:
+        case MAV_CMD_DO_SET_PARAMETER:
+        case MAV_CMD_DO_SET_RELAY:
+        case MAV_CMD_DO_REPEAT_RELAY:
+        case MAV_CMD_DO_SET_SERVO:
+        case MAV_CMD_DO_REPEAT_SERVO:
+        case MAV_CMD_DO_SET_REVERSE:
+        case MAV_CMD_DO_CONTROL_VIDEO:
+        case MAV_CMD_DO_DIGICAM_CONFIGURE:
+        case MAV_CMD_DO_SET_CAM_TRIGG_DIST:
+        case MAV_CMD_DO_FENCE_ENABLE:
+        case MAV_CMD_NAV_SET_YAW_SPEED:
+        case MAV_CMD_DO_SET_CAM_TRIGG_INTERVAL:
+        case MAV_CMD_PREFLIGHT_CALIBRATION:
+        case MAV_CMD_SET_CAMERA_MODE:
+        case MAV_CMD_IMAGE_START_CAPTURE:
+        case MAV_CMD_IMAGE_STOP_CAPTURE:
+        case MAV_CMD_DO_TRIGGER_CONTROL:
+        case MAV_CMD_VIDEO_START_CAPTURE:
+        case MAV_CMD_VIDEO_STOP_CAPTURE:
+        case MAV_CMD_ARM_AUTHORIZATION_REQUEST:
+        case MAV_CMD_NAV_RALLY_POINT:
```

Comment:
This seems wrong, we don't currently support the cmd type at all, and it wouldn't show up in the actual mission either way...

Reply:
I've trimmed this list down.

@WickedShell and I discussed creating something like an `AP_NAV_CMD` enumeration which maps from the `MAV_CMD` enumeration into those ArduPilot supports.  This would allow us to eliminate several `default:` clauses and help ensure correct handling of new enumeration values.